### PR TITLE
Add unit test

### PR DESCRIPTION
**Pull request by a-dastyar**
Add unit test, bug fix and refactoring.

### CODE REVIEW

File: src/main/java/com/campus/banking/service/BankAccountServiceImpl.java [hunk 1]
```diff
@@ -16,9 +16,11 @@ public void deposit(BankAccount account, double amount) {
 
     @Override
     public void withdraw(BankAccount account, double amount) {
-        if (amount < 0 || amount > account.getBalance()) {
-            // TODO: divide to two if statement
-            throw new InsufficientFundsException("Can not withdraw negative amount");
+        if (amount < 0) {
+            throw new IllegalArgumentException("Can not deposit negative amount");
```
Comment (AmentBRT) [inline src/main/java/com/campus/banking/service/BankAccountServiceImpl.java:20]: typo on "deposit"

File: src/main/java/com/campus/banking/service/CheckingAccountServiceImpl.java [hunk 2]
```diff
@@ -2,75 +2,93 @@
...
+            throw new InvalidAccountTypeException("BankAccount type must be from type CheckingAccount");
+        }
+    }
 
-        super.deposit(checkingAccount, amount);
+    private void doDeposit(CheckingAccount account, double amount) {
+        var debt = account.getDebt();
+        var balance = account.getBalance();
+        if (account.getDebt() > 0.0d) {
+            if (account.getDebt() >= amount) {
```
Comment (AmentBRT) [inline src/main/java/com/campus/banking/service/CheckingAccountServiceImpl.java:31]: var "debt" can be used

File: src/main/java/com/campus/banking/service/CheckingAccountServiceImpl.java [hunk 3]
```diff
@@ -2,75 +2,93 @@
...
-                checkingAccount.setDebt(0.0d);
-                amount -= debt;
-            } else {
-                checkingAccount.setDebt(debt - amount);
-                amount = 0.0d;
-            }
+    private void doWithdraw(CheckingAccount account, double amount) {
+        var debt = account.getDebt();
+        var balance = account.getBalance();
+        if (amount > account.getBalance()) {
```
Comment (AmentBRT) [inline src/main/java/com/campus/banking/service/CheckingAccountServiceImpl.java:72]: var "balance" can be used

File: src/main/java/com/campus/banking/service/SavingAccountServiceImpl.java [hunk 4]
```diff
@@ -26,6 +26,6 @@ public void withdraw(BankAccount account, double amount) {
     public void applyInterest(SavingAccount account) {
         double interest = account.getBalance() * SavingAccount.INTEREST_RATE;
 
-        deposit(account, interest);
+        super.deposit(account, interest);
```
Comment (AmentBRT) [inline src/main/java/com/campus/banking/service/SavingAccountServiceImpl.java:29]: May implement deposit later on

File: src/test/java/com/campus/banking/persistence/DatabaseTest.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            db.persist();
+            db.load();
+            assertThat(db.list().size()).isEqualTo(2);
+        } catch (LoadFailureException | SaveFailureException e) {
+            fail();
+        }
+    }
+
+    @Test
+    void persist_withMultipleBankAccount_shouldSaveToFile() {
```
Comment (AmentBRT) [inline src/test/java/com/campus/banking/persistence/DatabaseTest.java:145]: typo on "Multiple"

File: src/test/java/com/campus/banking/persistence/DatabaseTest.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            db.persist();
+            db.load();
+            assertThat(db.list()).isEmpty();
+        } catch (LoadFailureException | SaveFailureException e) {
+            fail();
+        }
+    }
+
+    @Test
+    void persist_withOneBankAccount_shouldSaveToFile() {
```
Comment (AmentBRT) [inline src/test/java/com/campus/banking/persistence/DatabaseTest.java:124]: typo on "One"

Comment (a-dastyar) [reply src/test/java/com/campus/banking/persistence/DatabaseTest.java:124]: what is the correct spelling?

File: src/test/java/com/campus/banking/persistence/DatabaseTest.java (hunk 5 above)
Comment (AmentBRT) [reply src/test/java/com/campus/banking/persistence/DatabaseTest.java:145]: This function checks the "One"

Comment (a-dastyar) [reply src/test/java/com/campus/banking/persistence/DatabaseTest.java:145]: oh, right 😂

File: src/test/java/com/campus/banking/persistence/DatabaseTest.java (hunk 6 above)
Comment (a-dastyar) [reply src/test/java/com/campus/banking/persistence/DatabaseTest.java]: fixed it, thanks.

Comment (AmentBRT) [reply src/test/java/com/campus/banking/persistence/DatabaseTest.java]: No problem

**Review by AmentBRT — APPROVED** [review]
All done.

**Pull request merged**